Fix undefined method 'method name' for Rails/FindEach

Knack opened the pull request:
Fixes #6206

#6198 removed the generic #method_name matcher from Node, and #6249 fixed the Rails/FindEach cop.

However, there was still an error when processing the node ancestors. I've written a regression test and implemented a simple fix.
  
-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

wata727 commented:
Thanks for your work! This implementation seems to be able to fix the reported issue.

BTW, I think there is a problem with using ancestor nodes to retrieve method chains. For example, the AST of the following code is as follows.

```ruby
User.all.select(:name, :age).each { |u| u.something }
```

```
(block
  (send
    (send
      (send
        (const nil :User) :select
        (sym :name)
        (sym :age)) :all) :each)
  (args
    (arg :u))
  (send
    (lvar :u) :something))
```

Looking at the above, the method chain is included in the receiver. These aren't ancestor nodes but descendant nodes.

I'm a little worried whether we should fix this problem in this pull request. What do you think?

Knack commented:
Well, I don't know much about AST. I simply debugged a bit and applied the fix that seemed most logical to me. I also found the values of `ancestors` strange when debugging.

Maybe it would be better to solve the problem in another pull request, because this bug is present in version 0.59.0.

bbatsov commented:
I think it will better to fix the approach used to find the method chain, as this seems like a deeper problem to me.

Knack commented:
You were right, `method_chain` seemed not to work as expected. I added one more test that failed and implemented the method as I think it should work. Please see commit 2d282b7 ♻️

wata727 commented on lib/rubocop/cop/rails/find_each.rb:
```diff
@@ -39,7 +39,13 @@ def autocorrect(node)
+            chain << node.method_name
+            node = node.receiver
+            break unless node.send_type?
+          end
```
I think `Node#descendants` is better in this case :)

Knack replied:
Right! it looks like simply changing  `ancestors` by `descendants` on my first implementation will work. I've pushed another commit. Thanks!

wata727 commented on lib/rubocop/cop/rails/find_each.rb:
```diff
@@ -39,7 +39,7 @@ def autocorrect(node)
 
         def method_chain(node)
-          [*node.ancestors, node].map(&:method_name)
+          [*node.descendants.take_while(&:send_type?), node].map(&:method_name)
```
I would like to know why you choose `Enumerable#take_while`. For example, in the following cases, the behavior differs between `select` and` take_while`.

```ruby
User.order(:name).tap{ |foo| puts foo }.all.each { |u| u.something }
```

```
$ pry -rparser -rparser/ruby25 -rrubocop
[1] pry(main)> buffer = Parser::Source::Buffer.new('(string)')
[2] pry(main)> buffer.source = "User.order(:name).tap{ |foo| puts foo }.all.each { |u| u.something }"
[3] pry(main)> builder = RuboCop::AST::Builder.new
[4] pry(main)> parser = Parser::Ruby25.new(builder)
[5] pry(main)> node = parser.parse(buffer).send_node
[6] pry(main)> node.descendants.select(&:send_type?).map(&:method_name)
=> [:all, :tap, :order, :puts]
[7] pry(main)> node.descendants.take_while(&:send_type?).map(&:method_name)
=> [:all]
```

If use `take_while`, it cannot pick up all method chains, but if use `select`, it pick up the wrong `send` node in the block. It is difficult which is more desirable, but in terms of avoiding false positives, I guess `select` seems more desirable. WDYT?

Knack replied:
Yes, I think that both methods have false positives. For example, something like:

```ruby
User.where(id: Admin.select(:id)).all.each(&:something)
```

wouldn't register an offense because it will think that the `select` method for `Admin` is part of the method chain.

```
node.descendants.select(&:send_type?).map(&:method_name)
=> [:all, :where, :select]
node.descendants.take_while(&:send_type?).map(&:method_name).inspect
=> [:all, :where]
```

I really don't know which is more desirable.

Knack replied:
If we want to avoid false _positives_ then yes, it seems that `select` is safer. We can miss some real offenses but we won't get offenses that shouldn't be reported :+1:
